opm/autodiff/AquiferCarterTracy.hpp

```diff
@@ -0,0 +1,418 @@
+                        else if (faceTag == 5) // top
+                            faceDirection = Opm::FaceDir::ZPlus;
+
+                        if (faceDirection == connection.reservoir_face_dir.at(idx))
+                        {
+                            // Check now if the face is outside of the reservoir, or if it adjoins an inactive cell
+                            // Do not make the connection if the product of the two cellIdx > 0. This is because the
+                            // face is within the reservoir/not connected to boundary. (We still have yet to check for inactive cell adjoining)
+                            faceArea = (faceCells(faceIdx,0)*faceCells(faceIdx,1) > 0)? 0. : Opm::UgGridHelpers::faceArea(ugrid, faceIdx);
+                            faceArea_connected_.at(idx) = faceArea;
+                            denom_face_areas += ( connection.influx_multiplier.at(idx) * faceArea_connected_.at(idx) );
+                        }
+                    }
+                    auto cellCenter = grid.getCellCenter(cell_idx_.at(idx));
+                    cell_depth_.at(idx) = cellCenter[2];
+                }
+
+                for (size_t idx = 0; idx < cell_idx_.size(); ++idx)
+                {
+                    alphai_.at(idx) = ( connection.influx_multiplier.at(idx) * faceArea_connected_.at(idx) )/denom_face_areas;
+                }
+            }
+
+            inline void calculate_aquifer_condition()
+            {
+
+                int pvttableIdx = pvttableID_ - 1;
+                
+                rhow_.resize(cell_idx_.size(),0.);
+                
+                if (p0_defaulted_)
+                {
+                   pa0_ = calculate_reservoir_equilibrium(rhow_);
+                }
+
+                // Initialize a FluidState object first
+                FluidState fs_aquifer;
+                // We use the temperature of the first cell connected to the aquifer
+                // Here we copy the fluidstate of the first cell, so we do not accidentally mess up the reservoir fs
+                fs_aquifer.assign( ebos_simulator_.model().cachedIntensiveQuantities(cell_idx_.at(0), /*timeIdx=*/ 0)->fluidState() );
+                Eval temperature_aq, pa0_mean;
+                temperature_aq = fs_aquifer.temperature(0);
+                pa0_mean = pa0_;
+
+                Eval mu_w_aquifer = FluidSystem::waterPvt().viscosity(pvttableIdx, temperature_aq, pa0_mean);
+
+                mu_w_ = mu_w_aquifer.value();
+
+            }
+
+            // This function is for calculating the aquifer properties from equilibrium state with the reservoir
+            inline Scalar calculate_reservoir_equilibrium(std::vector<Eval>& rho_water_reservoir)
+            {
+                // Since the global_indices are the reservoir index, we just need to extract the fluidstate at those indices
+                std::vector<Scalar> water_pressure_reservoir, pw_aquifer;
```

Comment:
`water_pressure_reservoir` does not need to be a vector here. It can be one value for each `for` step.